### ISSUE TITLE
Ability to know which subqueue the value was pulled from?

### ISSUE THREAD

**Issue by 243826** [marianobarrios/linked-blocking-multi-queue#4]
First off all, thanks for writing this. This is almost what I am looking for.

The subject says what part is missing. Ideally I would also like to have an ability to store different data type on each queue. Although I can declare ```E``` to be ```Object``` and achieve it, it forces me to do the casting at later point.

Without an ability to know which queue it came from ```instanceof``` is the only alternative and it could lead to dangerous bugs (when for some values one type is subtype of other) which I would like to avoid.

Did you think of providing such a functionality before?

**Comment by varad** [comment]
As far as I understand, you want to put completely unrelated objects (no common ancestor, no common interface) to the queue. Then you wanna receive them from the queue and process each of the objects differently.

I believe you can achieve it even without modifying LBMQ.

You can create a wrapper class and store it it in the queue.

`public class Wrapper {`
`  Object obj; // your object`
`  Class<?> clazz; // data type of your object`
`}`

When you receive the wrapper from the queue you can do something like:

`if (wrapper.clazz == MyClass.class) {`
`  // do something`
`}`

**Comment by 243826** [comment]
Yes, that's a possible approach. But the disadvantage is the garbage it creates due to use of wrapper. In fact I even started considering LBMQ is because I did not want to create Runnable objects as they add to the garbage.

with single queue, currently I do this:

```
queue.add(new Runnable() { public void run() { processType1(type1Object); }});
queue.add(new Runnable() { public void run() { processType2(type2Object); }});

// somewhere else
queue.get().run();
```

Would be much better if I did the following:
```
queue.get(TYPE1).add(type1Object);
queue.get(TYPE2).add(type2Object);

// somewhere else:
{object, type}  = queue.get(); // this one is pseudo syntax; we will need to map to the efficient interface
switch (type) {
  case TYPE1:
       processType1(object);
       break;

  case TYPE2:
       processType2(object);
       break;
}

**Comment by marianobarrios** [comment]
> `{object, type}  = queue.get(); // this one is pseudo syntax; we will need to map to the efficient interface`

But how could the queue make this more efficient? We would need to add overrides that return some kind of container object. Those container objects will cause the same garbage.

**Comment by 243826** [comment]
as an example

```type = queue.get(objectHolder)```
 or
```object = queue.get(typeHolder);```



[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
06/23/18,
1:06:10 AM

**Comment by 243826** [comment]
as an example

```type = queue.get(objectHolder);```
 or
```object = queue.get(typeHolder);```

**Comment by marianobarrios** [comment]
Are we really sure that the extra allocation is not acceptable? This is a linked list, that, by design, allocates one object per node in the queue, so we already have significant allocation. Are you sure that you really have a use case that can tolerate the GC pressure generated by the list itself but not the additional pressure generated by the approach suggested that @varad?

Even if that was the case in one particular example (which should be proved with a test), I honestly still have concerns about adding such an unidiomatic and invasive change, just for one case that is unlikely to happen again.

**Comment by 243826** [comment]
I actually refractored and did away with multi queue because that entailed
writing my own executor service. Instead I am using the queue provided by
executor service with and focusing  on maanaging the garbage created by
Runnable instances.

So I personally don't need what I originally requested. Thanks for your
attention.
